10240 restore 'all-successful' CI job and simplify workflows files

graingert opened the pull request:
## Scope and purpose

we need a job to [select as a required build](https://github.community/t/is-it-possible-to-require-all-github-actions-tasks-to-pass-without-enumerating-them/117957/4?u=graingert) so that github "automerge" doesn't just instantly merge the PR

This also adds  `--warnings-as-errors` to pydcotor to prevent https://github.com/twisted/twisted/pull/1641 happening again


## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/10240
* [ ] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] The title of the PR starts with the associated Trac ticket number (without the `#` character).
* [x] I have updated the automated tests and checked that all checks for the PR are green.
* [ ] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.
* [ ] The merge commit will use the below format
  The first line is automatically generated by GitHub based on PR ID and branch name.
  The other lines generated by GitHub should be replaced.

```
Merge pull request #123 from twisted/4356-branch-name-with-trac-id

Author: <github_username>, <github_usernames_if_more_authors>
Reviewer: <github_username>, <github_username_if_more_reviewers>
Fixes: ticket:<trac_ticket_number>, ticket:<another_if_more_in_one_PR>

Long description providing a summary of these changes.
(as long as you wish)
```

graingert committed:
combine lint, mypy and docs workflows

adiroiban commented on .github/workflows/lint.yaml:
```diff
@@ -28,7 +28,7 @@ jobs:
       run: |
         python --version
-        tox -e lint
+        tox -e lint,apidocs
```
I think that instead of updating lint, we should fix the docs.yaml workflow to fail on pydoctor warnings

graingert replied:
we already have way too many workflows, I'd like to push more complexity into tox and out of all the .github files

adiroiban replied:
well...the docs.yaml just calls `tox -e apidocs` which calls `bin/admin/build-apidocs`  which calls `twisted.python._release.BuildAPIDocsScript` ... is just that for narrative docs and apidocs we have a separate workflow... so that for example if a flaky test fails we can re-run only the trial tests...without having to rerun lint and documentation generation

but yes.. I would say that mypy.yaml and docs.yaml and lint.yaml can be merged.

adiroiban replied:
I see that lint now takes 7 minutes... I think that one of the reason why docs was moved to a separate workflow, was to have the docs executed in a separate VM... so that we can get a fast feedback.

But maybe we can run tox on 2 CPUs ... or build the docs as a separate job inside the lint.yaml

graingert committed:
enable warnings-as-errors

graingert committed:
run lint workflow on py3.9

graingert committed:
add all-successful job

currently the required builds are broken, so automerge is also broken
with this all-successful job the required builds can be updated to only
require "all-successful" which automatically requires all other
jobs

graingert committed:
run lint jobs in parallel

disable tox parallel spinner
show output for tox parallel

graingert committed:
newsfragment

adiroiban commented on .github/workflows/test.yaml:
```diff
@@ -168,3 +168,49 @@ jobs:
+      run: |
+        python -m pip install --upgrade pip tox
+
+    - name: Lint
```
I think the name of the step (and also for the overall job) is no longer `lint` ... but I don't know what name to use :)
maybe "tox run" ?

graingert replied:
It's stuff that's `not test`

graingert replied:
What about calling it `nit`?

adiroiban replied:
I think that any name, other than 'lint' would be ok :) ... but this is the tox environments run ... 
but nit is a bit gross :p

graingert replied:
> this is the tox environments run 

the other stages run tox environments - specifically the `py` environments

adiroiban replied:
I don't have a better name... but I think that this is a good description for the step

```suggestion
    - name: Run the checks
```

adiroiban commented on .github/workflows/test.yaml:
```diff
@@ -168,3 +168,49 @@ jobs:
       with:
         password: ${{ secrets.PYPI_UPLOAD_TOKEN }}
+
+  lint:
```
mabye 
```suggestion
  StaticChecks:
```

graingert committed:
Update .github/workflows/test.yaml

Co-authored-by: Adi Roiban <adiroiban@gmail.com>

graingert commented on .github/workflows/test.yaml:
```diff
@@ -168,3 +168,49 @@ jobs:
+  all-successful:
+    # https://github.community/t/is-it-possible-to-require-all-github-actions-tasks-to-pass-without-enumerating-them/117957/4?u=graingert
+    runs-on: ubuntu-latest
+    needs: [testing, lint, pypi-publish]
```
```suggestion
    needs: [testing, StaticChecks, pypi-publish]
```

graingert committed:
Update .github/workflows/test.yaml

adiroiban reviewed (approved):
It looks good. Thanks.

Just a minor comment to update the name of the step, which is now called "lint" but it also runs many other checks.

graingert committed:
Update .github/workflows/test.yaml

Co-authored-by: Adi Roiban <adiroiban@gmail.com>
